### PR TITLE
fix: escape `%` if URI malformed

### LINKED ISSUE #5441: Generating a chunk with a name containing `%` will lead to `URI malformed` error
### Rollup Version

4.13.0

### Operating System (or Browser)

Stackblitz

### Node Version (if applicable)

_No response_

### Link To Reproduction

https://stackblitz.com/edit/rollup-repro-jhg2my?file=dist%2Fmain.js

### Expected Behaviour

The output chunk name doesn't use `%`. Or the `import('./foo%25bar-hash.js')` is generated instead of `import('./foo%bar-hash.js')`.

### Actual Behaviour

The output chunk name doesn't uses `%`.

### Additional Information
This is similar to #4222. Maybe it makes sense to add `%` to the default `sanitizeFileName` like it was done in #4262.

### PR DESCRIPTION
**Pull request by baseballyama**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

-  resolves #5441

### Description

At first I thought of simply replacing `%` with `_`, but that would intentionally break all URL-encoded filenames. This is not a good idea, so I used `decodeURIComponent` to change `%` to `_` only if an abnormal `%` is present.

At least one major Node.js / browser has `decodeURIComponent`, but whenever it does not, `%` is replaced with `_`.

### CODE REVIEW

**Comment by lukastaegert** [comment]
Not sure I see the value of `decodeURIComponent` in the tests at the moment. I would have expected `foo%20bar` to become `foo bar`, but this did not happen. Also, the current test does not show how the dynamic import would be rendered, would it really work? I would feel better if we just add `_` to the list of forbidden characters.

File: src/utils/sanitizeFileName.ts [hunk 1]
```diff
@@ -2,12 +2,30 @@
...
+	let path = name.slice(driveLetter.length).replace(INVALID_CHAR_REGEX, '_');
 
-	// A `:` is only allowed as part of a windows drive letter (ex: C:\foo)
-	// Otherwise, avoid them because they can refer to NTFS alternate data streams.
-	return driveLetter + name.slice(driveLetter.length).replace(INVALID_CHAR_REGEX, '_');
+	// Replace % with _ if there is invalid escape sequence
+	if (path.includes('%')) {
+		if (decodeURIComponentSafe) {
+			try {
+				decodeURIComponentSafe(path);
```
Comment (lukastaegert) [inline src/utils/sanitizeFileName.ts:21]: Maybe you should have actually assigned the return value to `path`?

File: src/utils/sanitizeFileName.ts [hunk 2]
```diff
@@ -2,12 +2,30 @@
 // eslint-disable-next-line no-control-regex
 const INVALID_CHAR_REGEX = /[\u0000-\u001F"#$&*+,:;<=>?[\]^`{|}\u007F]/g;
 const DRIVE_LETTER_REGEX = /^[a-z]:/i;
+const PERCENT_REGEX = /%/g;
+
+const decodeURIComponentSafe =
+	typeof decodeURIComponent === 'undefined' ? undefined : decodeURIComponent;
```
Comment (TrickyPi) [inline src/utils/sanitizeFileName.ts:8]: Slight thought, `decodeURIComponent` has been widely supported. I think we can always assume it exists. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent#browser_compatibility

**Comment by baseballyama** [comment]
> I would have expected foo%20bar to become foo bar, but this did not happen.

I kept `foo%20bar` because if we convert `foo%20bar` to `foo bar`, it will be breaking change because all legal `%` that are not URI malformed will always be converted.

Therefore I replace only illegal `%` to `_`.

File: src/utils/sanitizeFileName.ts (hunk 2 above)
Comment (baseballyama) [reply src/utils/sanitizeFileName.ts:8]: Yes. All browsers and node / deno / bun are supported `decodeURIComponent`.
So I also think that we can directly use `decodeURIComponent`.
My initial implementation is super conservative.

File: src/utils/sanitizeFileName.ts (hunk 1 above)
Comment (baseballyama) [reply src/utils/sanitizeFileName.ts:21]: I wrote reason why I didn't return the return value.

https://github.com/rollup/rollup/pull/5535#issuecomment-2141331133

File: src/utils/sanitizeFileName.ts (hunk 2 above)
Comment (baseballyama) [reply src/utils/sanitizeFileName.ts]: updated: https://github.com/rollup/rollup/pull/5535/commits/38113286fb17fbe9300efa6609019c9fc0d0c151

**Comment by baseballyama** [comment]
> Also, the current test does not show how the dynamic import would be rendered

I added tests.

**Comment by lukastaegert** [comment]
> I kept foo%20bar because if we convert foo%20bar to foo bar, it will be breaking change because all legal % always be converted.

This will not work if you not also replace the file name on disk. Take the original stackblitz reproduction https://stackblitz.com/edit/rollup-repro-jhg2my?file=dist%2Fmain.js

If you instead use %20 in the file name and build it and try to run it via `node dist/main.js` then it will still fail. The reason is apparently that the importer does first replace `%20` with ` ` and then looks for `foo bar.js` on the disk, which does not exist.

Either you replace it all the time with `_`, which I would find perfectly fine for me, or you actually use the output of `decodeURIComponent`. In that case, it would use the non-escaped name also in the dynamic import statement, which would also be ok as JavaScript does not require imports to be URL-encoded as far as I tried it out.

In any case, I do not think this is a breaking change either way because using `%` in file names was broken before anyway. Or to put it differently, it is impossible to import a file containing "%".

**Comment by baseballyama** [comment]
Thank you for checking this PR again!🙏

> This will not work if you not also replace the file name on disk. Take the original stackblitz reproduction

I think it will be fixed because decodeURIComponent('./foo%bar-q8VKMzZt.js') throws an exception and goes to the catch statement, replacing % with _.

https://github.com/rollup/rollup/pull/5535/files#diff-6d5ded6774a9966c5b0f5b61b0f8f9f49a4cc4864057dcdb5d8693dd9eac222bR22-R29

> If you instead use %20 in the file name and build it and try to run it via node dist/main.js then it will still fail. The reason is apparently that the importer does first replace %20 with   and then looks for foo bar.js on the disk, which does not exist.

This is the reason why I don’t use the return value of decodeURIComponent. I’m using a name even if % is contained as much as possible, but if the name cannot be used, I replace % with _.

> In any case, I do not think this is a breaking change either way because using % in file names was broken before anyway. Or to put it differently, it is impossible to import a file containing "%".

If the rollup team doesn’t recognize this as a breaking change, I will follow their decision. Obviously, just adding % to `INVALID_CHAR_REGEX` is much easier than my current changes.

I could be wrong (as this is my first PR in the rollup repo), but I don’t feel that there is a correct common understanding of the changes in this PR. So, for now, I will wait for your response instead of updating the PR.
(I'm not opposite about updating PR👍)

**Comment by lukastaegert** [comment]
> This is the reason why I don’t use the return value of decodeURIComponent. I’m using a name even if % is contained as much as possible

This is my point, to my understanding it is impossible to import a file that contains a percent sign EXCEPT if the import is written differently from the file name and URL-escapes the percent sign. E.g. if the file is named `foo%20bar.js`, the only way to import it is via `foo%2520bar`, encoding the percent sign as %25. At the moment, however, Rollup does not support writing the import different from the file name. Therefore, files on disk should be forbidden from using percent signs one way or another, at least that is my conclusion. But please try this out yourself.

**Comment by baseballyama** [comment]
Finally I got the point! Thank you for your explanation!
I updated the PR.

**Review by lukastaegert — APPROVED** [review]
Sorry for letting this drop for a while, of course this is fine now!

**Pull request merged**